### PR TITLE
Various small touch-ups for VM reservations

### PR DESCRIPTION
**Pull request by markjdb**
The commit log messages are hopefully sufficiently detailed. Verified using cheribsdtest with hybrid and purecap kernels.

Most of these are minor changes. The vmspace_fork() change is an important bug fix though. I'd be happy to split this into smaller PRs if that would be helpful.

### CODE REVIEW

File: sys/vm/vm_map.c [hunk 1]
```diff
@@ -2112,15 +2136,17 @@ vm_map_findspace(vm_map_t map, vm_offset_t start, vm_size_t length)
  */
 int
 vm_map_fixed(vm_map_t map, vm_object_t object, vm_ooffset_t offset,
-    vm_pointer_t start, vm_size_t length, vm_prot_t prot,
+    vm_pointer_t start, vm_pointer_t *cap, vm_size_t length, vm_prot_t prot,
```
[the remark this answers is not in the document]
Comment (markjdb) [reply sys/vm/vm_map.c]: Sorry, I missed this comment. I think it makes sense to 1) rename `addr` to `reservp` (this name is used elsewhere in the reservation code, albeit a bit differently) now that it's only used for returning a capability for the maybe-adjusted mapping, and 2) drop `MAP_CREATE_RESERVATION` and use `reservp != NULL` to indicate that a new reservation is to be created. I'd much rather not have both `MAP_RESERVATION_CREATE` and `MAP_CREATE_RESERVATION`.

**Comment by markjdb** [comment]
The latest push addresses the last unresolved issue. I also added a commit to fix error handling in `vm_map_reservation_delete()`, and removed a commit which asserted that the length passed to `vm_map_findspace()` is representable, as ASLR breaks that assumption when it adds a randomized gap to the size of the region passed to `vm_map_findspace()` in `vm_map_find()`.

**Comment by jrtc27** [comment]
Is that code path used for freebsd32/freebsd64? CheriABI doesn't enable ASLR as it's a waste of time when you have capabilities.

**Comment by markjdb** [comment]
> Is that code path used for freebsd32/freebsd64?

Yes, I hit this problem when running cheribsdtest-hybrid on a purecap kernel. Hmm, I think I see the real bug(?) now: when a process exec()s, exec_new_vmspace() will either set or clear MAP_RESERVATIONS depending on whether the image uses the CHERI ABI or not. But when a non-CHERI ABI process forks, _vm_map_init() will set MAP_RESERVATIONS again, and we end up with a process that has both MAP_ASLR and MAP_RESERVATIONS set. This behaviour exists only in purecap kernels.

I think it's undesirable to leave MAP_RESERVATIONS set in non-CHERI processes, so probably the right thing to do is fix that bug and restore the assertion I had added to vm_map_findspace().

> CheriABI doesn't enable ASLR as it's a waste of time when you have capabilities.

Indeed.

**Comment by markjdb** [comment]
Latest push restores an assertion in vm_map_findspace() and ensures that MAP_RESERVATIONS isn't leaked into non-CHERI processes.

**Review by bsdjhb — APPROVED** [review]
Couple of questions: 1) do we want to axe the now-unused vm_map_reservation_delete() function?  2) Do we want to upstream MS_PAGEOUT?  I've had to fix it a few times during the last year's-worth of merges due to upstream API changes that we wouldn't have to do if it were upstream.

**Comment by markjdb** [comment]
> Couple of questions: 1) do we want to axe the now-unused vm_map_reservation_delete() function? 2) Do we want to upstream MS_PAGEOUT? I've had to fix it a few times during the last year's-worth of merges due to upstream API changes that we wouldn't have to do if it were upstream.

I think it's fine to axe vm_map_reservation_delete(). I'll push a commit to remove it.

I'd like to have MS_PAGEOUT upstream if only for testing purposes. I'd want to think a bit about whether we want to permit it for unprivileged users. I guess we already have MS_SYNC doing the same thing for filesystem objects, so maybe not. And presumably you could trigger pageouts to swap with msync(tmpfs file) anyway. I'll bring it up with Kostik, I think he implemented it originally.

**Comment by bsdjhb** [comment]
Yes, MS_PAGEOUT was a patch Kostik originally.

**Pull request merged**